mithril-common/src/digesters/cardano_immutable_digester.rs

```diff
@@ -49,99 +84,125 @@ impl ImmutableDigester for CardanoImmutableDigester {
+
+        let leaves: Vec<_> = computed_immutables_digests
+            .digests
+            .into_iter()
+            .map(|digest| MKTreeNode::new(digest.into()))
+            .collect();
+
+        let mktree =
+            MKTree::new(&leaves).map_err(ImmutableDigesterError::MerkleTreeComputationError)?;
+
+        debug!(
+            self.logger,
+            "Successfully computed Merkle tree for Cardano database"; "beacon" => #?beacon);
+
+        if let Some(cache_provider) = self.cache_provider.as_ref() {
+            if let Err(error) = cache_provider
+                .store(computed_immutables_digests.new_cached_entries)
+                .await
+            {
+                warn!(
+                    self.logger, "Error while storing new immutable files digests to cache";
+                    "error" => ?error
+                );
             }
         }
+
+        Ok(mktree)
+    }
+}
+
+fn list_immutable_files_to_process(
+    dirpath: &Path,
+    up_to_file_number: ImmutableFileNumber,
+) -> Result<Vec<ImmutableFile>, ImmutableDigesterError> {
+    let immutables: Vec<ImmutableFile> = ImmutableFile::list_completed_in_dir(dirpath)?
+        .into_iter()
+        .filter(|f| f.number <= up_to_file_number)
+        .collect();
+
+    match immutables.last() {
+        None => Err(ImmutableDigesterError::NotEnoughImmutable {
+            expected_number: up_to_file_number,
+            found_number: None,
+            db_dir: dirpath.to_owned(),
+        }),
+        Some(last_immutable_file) if last_immutable_file.number < up_to_file_number => {
+            Err(ImmutableDigesterError::NotEnoughImmutable {
+                expected_number: up_to_file_number,
+                found_number: Some(last_immutable_file.number),
+                db_dir: dirpath.to_owned(),
+            })
+        }
+        Some(_) => Ok(immutables),
     }
 }
 
-fn compute_hash(
+fn compute_immutables_digests(
     logger: Logger,
-    cardano_network: String,
-    beacon: &CardanoDbBeacon,
     entries: BTreeMap<ImmutableFile, Option<HexEncodedDigest>>,
-) -> CacheComputationResult {
-    let mut hasher = Sha256::new();
+) -> ComputedImmutablesDigestsResult {
     let mut new_cached_entries = Vec::new();
     let mut progress = Progress {
         index: 0,
         total: entries.len(),
     };
 
-    hasher.update(compute_beacon_hash(&cardano_network, beacon).as_bytes());
+    let mut digests = Vec::with_capacity(entries.len());
 
     for (ix, (entry, cache)) in entries.iter().enumerate() {
         match cache {
             None => {
                 let data = hex::encode(entry.compute_raw_hash::<Sha256>()?);
-                hasher.update(&data);
+                digests.push(data.clone());
                 new_cached_entries.push((entry.filename.clone(), data));
             }
             Some(digest) => {
-                hasher.update(digest);
+                digests.push(digest.to_string());
             }
         };
```

Comment:
This could maybe be written in a more functional way?

Reply:
WDYT about this implementation:

```
fn compute_immutables_digests(
    logger: Logger,
    entries: BTreeMap<ImmutableFile, Option<HexEncodedDigest>>,
) -> ComputedImmutablesDigestsResult {
    let mut new_cached_entries = Vec::new();
    let mut progress = Progress {
        index: 0,
        total: entries.len(),
    };

    let digests: Result<Vec<String>, io::Error> = entries
        .iter()
        .enumerate()
        .map(|(ix, (entry, cache))| {
            let digest = match cache {
                None => {
                    let data = hex::encode(entry.compute_raw_hash::<Sha256>()?);
                    new_cached_entries.push((entry.filename.clone(), data.clone()));
                    data
                }
                Some(digest) => digest.to_string(),
            };

            if progress.report(ix) {
                info!(logger, "Hashing: {progress}");
            }

            Ok(digest)
        })
        .collect();

    Ok(ComputedImmutablesDigests {
        digests: digests?,
        new_cached_entries,
    })
}
```

Reply:
The code above is outdated. It's better to refer to the comment below instead.